text/0192-remove-constructs-compat.md

```diff
@@ -0,0 +1,667 @@
+`Stage.of(this).outdir` to find the output directory of the current stage.
+
+### Implications on end-users
+
+Participation in synthesis is an "advanced" feature of the CDK framework and se
+assume most end-users don't use this directly.
+
+If they need "last minute processing", they would likely use `prepare()` (which
+is also being [TODO: removed]) but not `synthesize()`.
+
+The use case of emitting arbitrary files into the cloud assembly directory is
+weak. The cloud assembly is a well-defined format, and is currently "closed".
+There are no extension points that tools can identify. To that end, just writing
+files to the cloud assembly output directory does not make tons of sense. Having
+said that, it is still possible to do in the same way we plan for
+`AssetStaging`.
+
+
+### Using stacks as roots for unit tests
+
+If we move staging of assets to the initialization phase, it means we need to
+know at that time where is the cloud assembly output directory. As mentioned
+above, in production this information is available from the enclosing `Stage`.
+
+This works in production but introduces a minor issue with unit tests which use
+a `Stack` as the root construct. This is a very common pattern in the CDK
+codebase today:
+
+```ts
+const stack = new Stack();
+const foo = new MyFoo(stack, 'MyFoo');
+
+expect(stack).to(haveResource('AWS::Foo'));
+```
+
+In such cases, assets will not work because the output directory is only
+determined later (in the `expect` call).
+
+One alternative would be to simply modify these unit tests so that stacks are no
+longer used as roots (basically add `new App()` as the scope). This approach
+would require a change in many unit tests across the code base with no clear
+value to users.
+
+We propose to modify the `Stack` construct such that if a stack is created
+without an explicit `scope`, an `App` instance will automatically be created and
+used as it's scope.
+
+```ts
+const stack = new Stack();
+assert(stack.node.scope instanceof App); // previously it was `undefined`
+```
+
+Since only the root construct may have an empty ID, we will also need to assign
+an ID. We propose to use `"Stack"` since we already have fallback logic that
+uses this as the stack name when the stack does not have an ID (see [TODO]).
+
+This change will allow us to remove any special casing we have for stacks in the
+testing framework and throughout the synthesis code path (we have quite a lot of
+that), because we will be able to assume that `Stage.of(construct.node.root)` is
+never `undefined` and has a `synth()` method which returns a cloud assembly.
+
+Since unit tests sometimes use "incremental tests" for synthesized templates,
+and `stage.synth()` would reuse the synthesized output if called twice, we will
+also need to introduce a `stage.synth({ forceResynth: true })` option. This will
+be the default behavior when using `expect(stack)` or `SynthUtils.synth()`.
```

Comment:
Not clear why. If I `expect(stack)` multiple times on the same stack, I most likely don't expect it to be synthesized over and over again.

For `SynthUtils.synt()` I'd say it likely makes sense (since in this case, the user clearly expressed they intention to synthesize).

Reply:
This is the current behavior (re-synth) and I rather try to avoid breaking additional tests. It's mainly a performance optimization.